src/SFML/Window/Unix/KeyboardImpl.cpp

```diff
@@ -0,0 +1,818 @@
+
+    KeySym keysym = NoSymbol;
+    KeyCode keycode = scancodeToKeyCode(code);
+
+    if (keycode != NullKeyCode) // ensure that this Scancode is mapped to keycode
+        keysym = XkbKeycodeToKeysym(display, keycode, 0, 0);
+
+    sf::priv::CloseDisplay(display);
+
+    return keysym;
+}
+
+
+////////////////////////////////////////////////////////////
+bool isKeyPressedImpl(KeyCode keycode)
+{
+    if (keycode != NullKeyCode)
+    {
+        Display* display = sf::priv::OpenDisplay();
+
+        // Get the whole keyboard state
+        char keys[32];
+        XQueryKeymap(display, keys);
+
+        sf::priv::CloseDisplay(display);
+
+        // Check our keycode
+        return (keys[keycode / 8] & (1 << (keycode % 8))) != 0;
+    }
+
+    return false;
+}
+
+} // anonymous namespace
+
+namespace sf
+{
+namespace priv
+{
+
+////////////////////////////////////////////////////////////
+bool KeyboardImpl::isKeyPressed(Keyboard::Key key)
+{
+    KeyCode keycode = keyToKeyCode(key);
+    return isKeyPressedImpl(keycode);
+}
+
+
+////////////////////////////////////////////////////////////
+bool KeyboardImpl::isKeyPressed(Keyboard::Scancode code)
+{
+    KeyCode keycode = scancodeToKeyCode(code);
+    return isKeyPressedImpl(keycode);
+}
+
+
+////////////////////////////////////////////////////////////
+Keyboard::Scancode KeyboardImpl::delocalize(Keyboard::Key key)
+{
+    KeyCode keycode = keyToKeyCode(key);
+    return keyCodeToScancode(keycode);
+}
+
+
+////////////////////////////////////////////////////////////
+Keyboard::Key KeyboardImpl::localize(Keyboard::Scancode code)
+{
+    KeySym keysym = scancodeToKeySym(code);
+    return keySymToKey(keysym);
+}
+
+
+////////////////////////////////////////////////////////////
+String KeyboardImpl::getDescription(Keyboard::Scancode code)
+{
+    bool checkInput = true;
+
+    // these scancodes actually correspond to keys with input
+    // but we want to return their description, not their behaviour
+    if (code == Keyboard::Scan::Enter          ||
+        code == Keyboard::Scan::Escape         ||
+        code == Keyboard::Scan::Backspace      ||
+        code == Keyboard::Scan::Tab            ||
+        code == Keyboard::Scan::Space          ||
+        code == Keyboard::Scan::ScrollLock     ||
+        code == Keyboard::Scan::Pause          ||
+        code == Keyboard::Scan::Delete         ||
+        code == Keyboard::Scan::NumpadDivide   ||
+        code == Keyboard::Scan::NumpadMultiply ||
+        code == Keyboard::Scan::NumpadMinus    ||
+        code == Keyboard::Scan::NumpadPlus     ||
+        code == Keyboard::Scan::NumpadEqual    ||
+        code == Keyboard::Scan::NumpadEnter    ||
+        code == Keyboard::Scan::NumpadDecimal)
+    {
+        checkInput = false;
+    }
+
+    if (checkInput)
+    {
+        KeySym keysym = scancodeToKeySym(code);
+        Uint32 unicode = keysymToUnicode(keysym);
+
+        if (unicode != 0)
+            return String(unicode);
+    }
```

Comment:
Should this all just be another `switch` statement?

Reply:
I don't know, does it really matter? With a switch, it will complain if you don't provide a default value or don't handle all the cases, but we're just interested in a subset. I don't exactly find it less readable.

Reply:
It doesn't really matter. I just wanted to ensure it was considered.